javalib/src/main/scala/java/util/Hashtable.scala

```diff
@@ -87,25 +87,96 @@ class Hashtable[K, V] private (inner: mutable.HashMap[Box[Any], V])
+    new AbstractSet[K] {
+      def iterator(): Iterator[K] =
+        new EntrySetIterator().scalaOps.map(_.getKey())
+
+      def size(): Int = inner.size
+    }
+  }
 
   def entrySet(): ju.Set[ju.Map.Entry[K, V]] = {
-    class UnboxedEntry(
-        private[UnboxedEntry] val boxedEntry: ju.Map.Entry[Box[Any], V])
-        extends ju.Map.Entry[K, V] {
-      def getKey(): K = boxedEntry.getKey.inner.asInstanceOf[K]
-      def getValue(): V = boxedEntry.getValue
-      def setValue(value: V): V = boxedEntry.setValue(value)
-      override def equals(o: Any): Boolean = o match {
-        case o: UnboxedEntry => boxedEntry.equals(o.boxedEntry)
-        case _               => false
+    new AbstractSet[Map.Entry[K, V]] {
+      def iterator(): Iterator[Map.Entry[K, V]] =
+        new EntrySetIterator
+
+      def size(): Int = inner.size
+    }
+  }
+
+  /* Inspired by the implementation of
+   * scala.collection.convert.JavaCollectionWrappers.MapWrapper.entrySet
+   * as found in version 2.13.0-RC1, with two changes:
+   *
+   * - accommodate the fact that our keys are boxed, and
+   * - explicitly snapshot the underlying contents right before any mutation of
+   *   the underlying Map, as we do not have any guarantee that mutations
+   *   preserve the state of existing iterators.
+   */
+  private class EntrySetIterator extends Iterator[Map.Entry[K, V]] {
+    private var underlying: scala.collection.Iterator[(Box[Any], V)] =
+      Hashtable.this.inner.iterator
+
+    private var isSnapshot: Boolean = false
+
+    private var prev: Box[Any] = null
+
+    private def ensureSnapshot(): Unit = {
+      if (!isSnapshot) {
+        underlying = underlying.toList.iterator
+        isSnapshot = true
+      }
+    }
+
+    def hasNext(): Boolean = underlying.hasNext
+
+    def next(): Map.Entry[K, V] = {
+      val (boxedKey, initialValue) = underlying.next()
+      prev = boxedKey
+
+      new Map.Entry[K, V] {
+        private var value = initialValue
+
+        def getKey(): K = boxedKey.inner.asInstanceOf[K]
+
+        def getValue(): V = value
+
+        def setValue(v: V): V = {
+          ensureSnapshot()
+          val oldValue = value
+          inner.put(boxedKey, v)
+          value = v
+          oldValue
+        }
+
+        override def equals(that: Any): Boolean = that match {
+          case that: Map.Entry[_, _] =>
+            getKey() === that.getKey() && getValue() === that.getValue()
+          case _ =>
+            false
+        }
+
+        override def hashCode(): Int =
+          boxedKey.hashCode ^ (if (value == null) 0 else value.hashCode())
       }
-      override def hashCode(): Int = boxedEntry.hashCode()
     }
-    inner.asJava.entrySet().asScala.map(new UnboxedEntry(_): ju.Map.Entry[K, V]).asJava
+
+    def remove(): Unit = {
+      if (prev == null)
+        throw new IllegalStateException("next must be called at least once before remove")
+      ensureSnapshot()
+      inner -= prev
+      prev = null
+    }
```

Comment:
It seems this codepath is completely untested or am I missing something?

Reply:
I added a test for the corner cases of the entrySet's iterator.